tests/lib/bin/session-tool

```diff
@@ -0,0 +1,110 @@
+#!/bin/sh -e
+if [ "$(id -u)" -ne 0 ]; then
+    echo "session-tool needs to be invoked as root" >&2
+    exit 1
+fi
+if [ -z "$(command -v busctl)" ]; then
+    echo "session-tool requires busctl" >&2
+    exit 1
+fi
+if [ $# -eq 0 ]; then
+	echo "usage: session-tool [-u USER] [--] <CMD>"
+	exit 1
+fi
+user=root
+while [ $# -gt 0 ]; do
+	case "$1" in
+		--)
+			shift
+			break
+			;;
+		-u)
+			if [ $# -eq 1 ]; then
+				echo "session-tool: option -u requires an argument" >&2
+				exit 1
+			fi
+			user="$2"
+			shift 2
+			;;
+		-*)
+			echo "session-tool: unsupported argument $1" >&2
+			exit 1
+			;;
+		*)
+			break
+			;;
+	esac
+done
+
+sed -i -e 's/mesg n .*true/tty -s \&\& mesg n/g' /root/.profile
+
+read -r uuid < /proc/sys/kernel/random/uuid
+unit_name="session-tool-$uuid.service"
+tmp_dir="/tmp/session-tool-$uuid"
+mkdir "$tmp_dir"
+mkfifo -m 0666 "$tmp_dir/dbus-monitor.pipe" "$tmp_dir/ready.pipe" "$tmp_dir/result.pipe" "$tmp_dir/stdin.pipe" "$tmp_dir/stdout.pipe" "$tmp_dir/stderr.pipe"
+trap 'rm -rf $tmp_dir' EXIT
+
+# Run dbus-monitor waiting for systemd JobRemoved signal. Process the output
+# with awk, forwarding the result of the service to a fifo.
+monitor_expr="type='signal', sender='org.freedesktop.systemd1', interface='org.freedesktop.systemd1.Manager', path='/org/freedesktop/systemd1', member='JobRemoved'"
+stdbuf -oL dbus-monitor --system --monitor "$monitor_expr" > "$tmp_dir/dbus-monitor.pipe" &
+dbus_monitor_pid=$!
+
+awk_expr="BEGIN {found=0; ready=0;} /member=NameAcquired/ { if (!ready) {ready=1; print ready > \"$tmp_dir/ready.pipe\"; close(\"$tmp_dir/ready.pipe\"); } } /   string \"$unit_name\"/ {found=1; print \"found service file\"; fflush(); next; }; /string \".*\"/ { if (found==1) {print \"found result\"; print \$2 > \"$tmp_dir/result.pipe\"; fflush(\"\"); close(\"$tmp_dir/result.pipe\"); exit; } }"
+awk -W interactive "$awk_expr" <"$tmp_dir/dbus-monitor.pipe" >"$tmp_dir/awk.log" 2>&1 &
+awk_pid=$!
+
+# Wait for dbus-monitor to start.
+cat "$tmp_dir/ready.pipe" >/dev/null
+
+# Use busctl to spawn a command. The command is wrapped in shell, runuser -l
+# and redirects to capture output. Sadly busctl doesn't support passing file
+# descriptors https://github.com/systemd/systemd/issues/14954 As a workaround
+# we pass a set of pipes. This is good for non-interactive work.
+cat >"$tmp_dir/stdin.pipe" &
+cat <"$tmp_dir/stdout.pipe" >&1 &
+cat <"$tmp_dir/stderr.pipe" >&2 &
+busctl \
+	--allow-interactive-authorization=no --quiet \
+	--system \
+	-- \
+	call \
+	 org.freedesktop.systemd1 \
+	/org/freedesktop/systemd1 \
+	 org.freedesktop.systemd1.Manager \
+	StartTransientUnit "ssa(sv)a(sa(sv))" \
+	"$unit_name" fail 4 \
+		Description s "session-tool running $* as $user" \
+		Type s oneshot \
+		Environment as 1 TERM=xterm-256color \
+		ExecStart "a(sasb)" 1 \
+			/bin/sh 3 /bin/sh -c "exec runuser -l $user - -c \"$*\" <$tmp_dir/stdin.pipe >$tmp_dir/stdout.pipe 2>$tmp_dir/stderr.pipe" false \
```

Comment:
are there issues with quoting here, specifically around `\"$*\"` ? i.e. does this work if I do something like

```
$ session-tool -u test -- /bin/sh -c 'echo "$*" '
$ session-tool -u test -- /bin/sh -c "echo \"$*\" "
$ session-tool -u test -- /bin/sh -c 'echo $* '
```
etc. ?

might be worth a couple of tests in the spread test here to ensure that quoting is working the way we want it to since there's a few layers here

Reply:
I know but they are unfixable.

Originally this line read

`/sbin/runuser $(expr $# + 5) /sbin/runuser -l "$user" - -c "$@"` but this surfaced a bug where I was unable to rely on systemd's offering of stream redirection.

I'll add a comment.